### PR TITLE
boards/cpu: add support form EFM32 and starter kits

### PR DESCRIPTION
**Pull request by basilfx**
This PR adds initial support for the EFM32 and EZR32 MCUs by Silicon Labs. It refers to issue #1123.

The first commit adds the targets, the second commit adds the boards. It uses the existing flash scripts and Cortex architecture. It does not modify any other files.

**Overview**

The EFM32 and EZR32 MCUs are supported, including the new [EFM32 Gemstone MCUs](https://www.silabs.com/Support%20Documents/TechnicalDocs/AN0918.pdf). All peripherals, except random number generator, are supported.

An overview<sup>1,2</sup>:

* EFM32 Gecko
* EFM32 Giant Gecko
* EFM32 Happy Gecko
* EFM32 Jade Gecko
* EFM32 Leopard Gecko
* EFM32 Pearl Gecko
* EFM32 Tiny Gecko
* EFM32 Wonder Gecko
* EFM32 Zero Gecko
* EZR32 Leopard Gecko

It targets the following boards<sup>2</sup>:

* STK3200 &mdash; EFM32 Zero Gecko
* SLSTK3401A &mdash; EFM32 Pearl Gecko
* STK3600 &mdash; EFM32 Leopard Gecko
* STK3700 &mdash; EFM32 Giant Gecko
* STK3800 &mdash; EFM32 Wonder Gecko

**Features**

* Low-power UART support (can be disabled via `LOW_POWER_ENABLED=0`, disables USB UART on STK3200, because it is connected to LEUART0).
* Code Correlation for Advanced Energy Monitor (Cortex M3/M4 only, can be disabled via `AEM_ENABLED=0`)
* RTT *or* RTC. As I don't know how #2538 will end, I implemented both. The new EFM32 Gemstone MCUs have proper RTC support, the older ones use time conversions. It can be configured as RTC via `RTT_AS_RTC=1`.

**Documentation**

I have prepared wiki pages for the boards [here](https://github.com/basilfx/EFM2Riot/tree/master/dist/doc).

**Testing**

I have a STK3200, SLSTK3401A and STK3600. The STK3700 and STK3800 are identical to the STK3600. 

The applicable peripheral tests compile and work, with a few exceptions:

* `periph_rtc` fails on STK3200 (problems with timer).
* `periph_uart` exceeds RAM on STK3200.
* `periph_spi` compiles and works, but I haven't found a way to actually verify what is send on the pins.

I noticed two odds:

* `periph_spi` hardcodes PULLUP for output pins. This is not supported by the GPIO peripheral of EFM32.
* `perip_rtc` behaviour depends on implementation (`time_t` wrapper vs dedicated peripheral).

**TODO**

* Optimizing code (I'm not a C expert).
* Fixing SPI.
* More testing.
* Benchmarking energy consumption.

Long term (probably not in this PR):
* Adding DMA support
* Adding support for LETIMER. I used the [EZR32WG](https://github.com/RIOT-OS/RIOT/blob/master/cpu/ezr32wg/periph/timer.c) timer implementation, but I don't like using two 16-bit timers to create a 32-bit timer. I don't know the timer requirements, but reducing it to 16-bit shortens the period.

**Conclusion**

I know this PR is using vendor libraries. As mentioned in #1123, it is not a board support package. I also know that RIOT-OS targets as efficient code as possible. That said, emlib (the vendor library) does increase the code size, but adds flexibility. I have reported this [before](https://github.com/RIOT-OS/RIOT/issues/1123#issuecomment-159417033), but I expected a constant increase. To verify/research this, I conducted several benchmarks with different compile options and a few boards. The results of approximately 2200 compilations can be found [here](http://basilfx.github.io/EFM2Riot/benchmark_20160129/index.html) (large file!)<sup>3,4</sup>. I see an increase of 3-4 kb compared to other boards<sup>5</sup>.

I fully understand that the above may influence the decision *not* to include this PR in RIOT-OS. But #1123 is open for quite some time, and this could serve as a starting point for [optimization](https://github.com/RIOT-OS/RIOT/issues/1123#issuecomment-159589863). The targeted MCUs have variants with enough flash and ram. Furthermore, It did not stop me from learning about RIOT-OS, EFM32, embedded and coding in general, so no hard feelings. I will probably still use it for my personal projects.

*Footnotes*

<sup>1</sup> EZR32 Wonder Gecko is missing, but is already included in RIOT-OS. There is no compatibility or overlap.

<sup>2</sup> SLWSTK6220A is missing, but is already included in RIOT-OS. There is no compatibility or overlap.

<sup>3</sup> Parallel builds using `make -jN` for `N > 1` failed in many cases with 'malformed archives'. Therefore a larger number of build failed. Manually cleaning and recompiling the same code for `N = 1` succeeded, however.

<sup>4</sup> Click the numbers in the TEXT, DATA or BSS column for more information.

<sup>5</sup> My code is verbose: I explicitly reset peripherals, and initialize all parameters. Furthermore, clock setup and frequency calculation is also responsible for a big part of code (but used by multiple peripherals).

### CODE REVIEW

File: boards/slstk3401a/board.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
```
Comment (kaspar030) [inline boards/slstk3401a/board.c]: freie universität berlin?

File: boards/slstk3401a/board.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+/**
+ * @ingroup     boards_slstk3401a
+ * @{
+ *
+ * @file
+ * @brief       Board specific implementations SLSTK3401A board
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
```
Comment (kaspar030) [inline boards/slstk3401a/board.c]: Is that you? ;)

Comment (basilfx) [inline boards/slstk3401a/board.c]: No, but I did re-use the same `board.c` from EZR32WG.

**Comment by haukepetersen** [comment]
Would it be possible to split up this PR? This PR adds >1500 files, which is not manageable for decent reviewing... I would propose you add each board (and their corresponding CPU) as separate PR, this should reduce the number of files at least to some extend...

**Comment by basilfx** [comment]
Yeah sure, I will do that. I think I will do that tomorrow.

**Comment by haukepetersen** [comment]
Thanks!

**Comment by basilfx** [comment]
I wasn't able to fix this last weekend, sorry. I found a few bugs while working on a Si7021 driver, which I want to fix first.

**Comment by haukepetersen** [comment]
sure, take your time!

**Comment by basilfx** [comment]
I will close this one in favor of #4824. Let's start with one CPU/Board and see where that goes :-)

**Pull request closed without merging**